std/assembly/array.ts

```diff
@@ -476,6 +476,61 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    assert(isArray<T>());
+    // Get the length and data start values
+    let length = this.length_;
+    let selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    let size: i32 = 0;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      size += child == null ? 1 : child.length;
```

Comment:
This could be optimized further with `<i32>load<u32>(changetype<usize>(child), offsetof<T>("length_")`. 

Will need to make another pass.